### DISCUSSION TITLE
Binary return from a Struct

### DISCUSSION THREAD

**Discussion by danmoser** [FreeOpcUa/python-opcua#1454] [Q&A] [answered]
Hi there,

I am trying to implement a feature I saw in a OPC-UA server that I could not find in python-asyncua. The goal is to have an object that returns the values of its children variables, so that the variable's values can be retrieved either directly, or grouped in this object.

I don't know the details of the OPC-UA server running that, but I could access it and got the following:
```
customObj = client.get_node(f"ns={idx_axis};s=customObj")

print("STRUCT", await customObj.read_value())
# Results:
STRUCT ExtensionObject(TypeId=NodeId(Identifier=101, NamespaceIndex=7, NodeIdType=<NodeIdType.Numeric: 2>), 
Body=b'\x00\x00\x00\x00\x00...')

print("STRUCT", await customObj.get_variables())
# Results:
STRUCT [Node(NodeId(Identifier='Var1', NamespaceIndex=7, NodeIdType=<NodeIdType.String: 3>)), 
Node(NodeId(Identifier='Var2', NamespaceIndex=7, NodeIdType=<NodeIdType.String: 3>)), ...] 
```
Any suggestions on how to mimic that?
Thanks in advance,
Daniel

**Comment by AndreasHeine** [comment]
sounds like ExposeItsArray https://reference.opcfoundation.org/v104/Core/docs/Part5/8.4.1/

**Reply by danmoser** [reply]
Thank you, Andreas! But do we have an example of it in python-asyncua repository?
I found a discussion here https://github.com/FreeOpcUa/python-opcua/issues/1110 , and it does not appear to be implemented...

**Reply by AndreasHeine** [reply]
no

**Comment by danmoser** [comment, marked as answer]
What about Binary Encoding? https://reference.opcfoundation.org/v104/Core/docs/Part6/7.4.4/
Do we have it on asyncua?

**Reply by schroeder-** [reply]
Asyncua does support BinaryEncoding, but not HTTPS transport.

**Reply by danmoser** [reply]
Thanks for the prompt response... I've checked the examples, and I could not find one for BinaryEncoding.
Any tips?

**Reply by schroeder-** [reply]
Asyncua only supports communicating via BinaryEncoding, for example every client call sends data as BinaryEncoding.

**Reply by AndreasHeine** [reply]
all are binary servers! its the only implemented encoding so far.

**Reply by danmoser** [reply]
Your are absolutely right, @AndreasHeine . I was actually referring to structured types https://reference.opcfoundation.org/v104/Core/docs/Part6/5.2.6/

**Reply by danmoser** [reply]
(please ignore the transport protocol I've pasted previously)

**Reply by AndreasHeine** [reply]
i guess you mean extensionobjects (structured datatypes)

https://reference.opcfoundation.org/v104/Core/docs/Part6/5.1.5/

**Reply by schroeder-** [reply]
Yes, they are convert form/to python dataclasses. For the server side see example examples/server-custom-structures-amd-enums.py

**Reply by danmoser** [reply]
@schroeder-  , unfortunately the structures example doesn't contain what I need. I want a structure to return the "Body" binary stream when calling `customObj.read_value()` -- or am I missing something?

**Reply by schroeder-** [reply]
```
import struct_to_binary from asyncua.ua.ua_binary.py
struct_to_binary(customObj)
```

**Reply by schroeder-** [reply]
Sry missed the read_value
```

import struct_to_binary from asyncua.ua.ua_binary.py
obj = await customObj.read_value()
struct_to_binary(obj)

```

**Reply by danmoser** [reply]
@schroeder- , that's great, thank you!
My final question: is there a way link the parent object struct with its children values (ie., a value update in a child will automatically be reflected as one of the struct entries)? Or is the proper way of doing that is using callbacks?

**Reply by schroeder-** [reply]
You have to link it your self and you can use callbacks to sync the values.

**Reply by danmoser** [reply]
Thank you!!
